edits for clarification and structure (#6)
zapatacomputing/orquestra-docs: intro/content/tutorial/helloworkflow.mdx

- Building a resource. A new resource will let you incorporate existing code into the workflow
- Writing templates that utilize resources to create in the `step` of a workflow

We could create a very minimalistic workflow without using resources, but we would miss out on what makes Quantum Engine powerful: sharing reusable templates and code. We will start by showing you how to build your own resource.

### Building a Resource

**1. Create a GitHub repository**

Go to [GitHub](https://github.com/) and create a public repository called `welcome-resource`. If you are unfamiliar with GitHub you can reference their [create a repo guide](https://help.github.com/en/github/getting-started-with-github/create-a-repo) for help

This repository will be where you build your resource. [This GitHub repo](https://github.com/zapatacomputing/tutorial-0-welcome) can be used as a reference for how `welcome-resource` should look like throughout the tutorial.

**2. Creating the Resource Folder Structure**

In order to be recognized by Orquestra, a resource must contain two folders:
- The `src/` folder which contains:

>   - A setup file that is responsible for installing your code in the machine where your workflow is executed. This installation is done automatically for you.

Reviewer: ```suggestion
  - A `setup.py` file that is responsible for installing your code in the machine where your workflow is executed. This installation is done automatically for you conforming to [Python3 setuptools standard](https://docs.python.org/3/distutils/setupscript.html) for Setup Scripts.
```

>   - Source code under a folder called `python/`.

Reviewer: ```suggestion
  - A nested folder `python/` containing source code.
```

> - The `templates` folder which will hold the templates usd by code in the `src` folder.

Reviewer: ```suggestion
- The `templates/` folder which will hold the templates used by code in the `src/` folder.
```

Using either the GitHub UI or by cloning your repo and using the command line create two folders at the root level of the repository, `src/` and `templates/`

**3. Adding `welcome.py` to `src/`**

Our goal in the first part of this tutorial is to create a resource that produces an [artifact](https://www.orquestra.io/docs/dcs/data/artifacts/) with the text `Welcome to Orquestra!` when used within a workflow. 

> We will start by adding some code to the `src/` folder. In general, Orquestra can support any Python source that is installable via `pip` using a `setup.py` script. 

Reviewer: ```suggestion
We will start by adding some code to the `src/` folder. In general, Orquestra can support any Python3 source that is installable via `pip` using a `setup.py` script. 
```

> Moving into the `src/` folder, create a folder structure `python/orquestra/` and with a python module in it called `welcome.py`. Your repo structure should now look like this:

Reviewer: ```suggestion
Moving into the `src/` folder, create a folder structure `python/orquestra/` and with a python script in it called `welcome.py`. Your repo structure should now look like this:
```

)
```

> In example, `name` is the name of your package. The contents of the `packages` field in this sample tells the installer to look in your repo in a folder called `python` for your package. The contents of the `package_dir` field in this sample means that the contents of the "python" folder can be imported without the python path prefix, for example, with `from orquestra import all`.

Reviewer: ```suggestion
In this example, `name` is the name of your package. The `packages` field in this example tells the installer to look in your repo in a folder called `python` for your package. The `package_dir` field in this sample means that the contents of the "python" folder can be imported without the python path prefix, for example, with `from orquestra import all`.
```

**6. Adding `templates`**

> Now that we've described the code that we want the resource to execute, we'll now work defining the `welcome` template for this resource. Move into the `templates/` folder and create a new file called `welcome.yaml`.

Reviewer: ```suggestion
Now that we've built the code that we want the resource to execute, we'll now work defining the `welcome` template for this resource. Move into the `templates/` folder and create a new file called `welcome.yaml`.
```

 
**Template**

> The `welcome.yaml` template is responsible for invoking the Python code that is living under `src/`. This can be done using the following snippet:

Reviewer: ```suggestion
The `welcome.yaml` template is responsible for invoking Python code that lives under `src/`. This can be done using the following snippet:
```

- The `outputs` field allows us to specify the data we expect at the end of the template's execution. Our Python script in input artfacts produced a file under `/app/welcome.json` but without output artifacts, we will have to look for the file by path all the time, and it could even be garbage collected since we never marked it as an output artifact. By specifying the file as an output artifact, we can refer to it as input artifacts in other templates.

**7. Push Your Resource**

> Having added our code under `src` and defined the templates, the final structure of the resource should look something like this:

Reviewer: ```suggestion
Having added our code under `src/` and defined our template under `templates/`, the final structure of the resource should resemble the following:
```

```
.
├── src
│   ├── python
│   │   └── orquestra
│   │       ├── __init__.py
│   │       └── welcome.py
│   └── setup.py
├── templates
│   └── welcome.yaml
```

> Once you are satisified with your work, commit your changes and push them back up to Github

Reviewer: ```suggestion
Once you are satisified with your work, commit your changes and push them to Github
```

**8. Building a Workflow**

> We can now build a simple workflow that uses the `welcome` resource to generate the welcome message artifact. Let's start by creating a `welcome-workflow.yaml` file with the following code:

Reviewer: ```suggestion
We can now build a simple workflow that uses the `welcome` git resource to generate the welcome message artifact. Let's start by creating a `welcome-workflow.yaml` file with the following code:
```

- The `templates` section in a workflow is where you describe, step by step, how to go from zero to your final results. You group these steps into templates so it is easier to work with them. Steps can run in parallel or in series. Series steps begin with a double dash, while parallel steps begin with a single dash. In our case, we only have a single step, so it begins with a double dash. The `name` of the template is how the `entrypoint`, discussed above, finds it. The first step, whose `name` is `welcome`, references the template `welcome-to-orquestra`, which we created earlier in our resource, using its `template` field.

**9. Adding Resources to the Workflow**

Next, we will see how to use a pre-existing resource and explore one possible pattern for a workflow with multiple steps.

> For example, Zapata has a resource with a URL of `git@github.com:zapatacomputing/tutorial-0-ztransform.git`. This will transform a message you give it in a fun way.

Reviewer: ```suggestion
For example, Zapata already has build a [transform message resource](https://github.com/zapatacomputing/tutorial-0-ztransform.git) with a git-URL of `https://github.com/zapatacomputing/tutorial-0-ztransform.git`. This resource translates a message you give it in a fun way.
```

- The template `z-transformation` takes in an artifact as input. The input is being fully-qualified accordingly. They must be placed inside a `{{ }}` placeholder.

**10. Running the Workflow**

You are now ready to run the workflow!

> * Make sure you have installed the [Quantum Engine CLI](http://localhost:8000/tutorial/hydrogen-vqe/#using-the-quantum-engine-cli)

Reviewer: ```suggestion
* Make sure you have installed the [Quantum Engine CLI](https://www.orquestra.io/docs/qe/cli/installing/)
```

* Log in to Quantum Engine by running `qe login -e <your-email> -s <quantum-engine-uri>` in your terminal. Contact support to register your email and/or receive the `quantum-engine-uri`.

* Submit your `welcome-workflow.yaml` by running `qe submit workflow <path/to/workflow/welcome-workflow.yaml>`**

This will return the workflow ID that corresponds to that particular execution of your workflow. The output will look like:
```Bash
Successfully submitted workflow to quantum engine!
Workflow ID: welcome-to-orquestra-d9djf
```

**11. Worfklow Progress**

> The workflow is now submitted to the Orquestra engine and will be scheduled for execution when compute becaomes available.

Reviewer: ```suggestion
The workflow is now submitted to the Orquestra Quantum Engine and will be scheduled for execution when compute becomes available.
```

Location:    http://40.89.251.200:9000/workflow-results/bb2b58b4-b25d-59e3-9fee-e7b79f0c20d5.json?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=zapata%2F20200319%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20200319T212017Z&X-Amz-Expires=604800&X-Amz-SignedHeaders=host&response-content-disposition=attachment%3B%20filename%3D%22bb2b58b4-b25d-59e3-9fee-e7b79f0c20d5.json%22&X-Amz-Signature=c4de1784b252fa6164aea8aa49a91bdd84c20c4dc55411e93f69a57b4ea62ac1
```

**13. Downloading the Results**

> When your workflow is completed, the `workflowresult` command will provide you with a click-able link under `Location` in the console output. Click on the link to download the file

Reviewer: ```suggestion
When your workflow is completed, the `workflowresult` command will provide you with a http web-link under `Location` in the console output. Click on or copy and paste the link into your browser to download the file
```
